crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+        );
+
+        Ok(ShrinkInfo {
+            output,
+            input: self.input,
+            input_size,
+            output_size: best.len() as u64,
+        })
+    }
+
+    fn is_valid_wasm(&self, file: &Path) -> Result<bool> {
+        let wasm =
+            std::fs::read(file).with_context(|| format!("Failed to read '{}'", file.display()))?;
+
+        let mut validator = wasmparser::Validator::new();
+        validator.wasm_features(wasmparser::WasmFeatures {
+            // TODO: we should have CLI flags for each Wasm proposal.
+            reference_types: true,
+            multi_value: true,
+            bulk_memory: true,
+            module_linking: true,
+            simd: true,
+            threads: true,
+            tail_call: true,
+            multi_memory: true,
+            exceptions: true,
+            memory64: true,
+
+            // We'll never enable this here.
+            deterministic_only: false,
+        });
+
+        Ok(validator.validate_all(&wasm).is_ok())
+    }
+
+    fn test_file(&self, file: &Path) -> Result<bool> {
+        let status = std::process::Command::new(&self.predicate)
+            .arg(file)
+            .stdin(std::process::Stdio::null())
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .status()
+            .with_context(|| {
+                format!(
+                    "Failed to spawn predicate script '{}'",
+                    self.predicate.display()
+                )
+            })?;
+
+        let code = status.code().expect(
+            "`wasm-shrink` doesn't work on platforms where child processes don't have exit codes",
+        );
+
+        Ok(code == 0)
+    }
+
+    fn test_bytes(&self, bytes: &[u8]) -> Result<bool> {
+        let file = new_temp_file()?;
+
+        std::fs::write(file.path(), bytes).with_context(|| {
+            format!(
+                "Failed to write to temporary file '{}'",
+                file.path().display()
+            )
+        })?;
+
+        self.test_file(file.path())
+    }
+}
+
+fn new_temp_file() -> Result<tempfile::NamedTempFile> {
+    let f = tempfile::NamedTempFile::new().context("Failed to create a temporary file.")?;
+    Ok(f)
+}
+
+fn read_file(path: &Path, into: &mut Vec<u8>) -> Result<()> {
+    use std::io::Read;
+    into.clear();
+    let mut file = std::fs::File::open(path)
+        .with_context(|| format!("Failed to open '{}'", path.display()))?;
+    file.read_to_end(into)
+        .with_context(|| format!("Failed to read '{}'", path.display()))?;
+    Ok(())
+}
+
+fn write_file(path: &Path, contents: &[u8]) -> Result<()> {
+    std::fs::write(path, contents)
+        .with_context(|| format!("Failed to write to the file '{}'", path.display()))?;
+    Ok(())
+}
+
+/// Information about a completed shrinking run.
+pub struct ShrinkInfo {
+    /// The path where the shrunken file was written to.
+    pub output: PathBuf,
```

Comment:
From a library perspective I think this would make more sense as `Vec<u8>`. Ideally the library version of this crate would have zero file operations baked in by default, and only the CLI embedding of the library would bring in the file operations

Reply:
Yeah this is just a bit harder to do with the builder-is-the-`structopt` pattern. Doable with a little nesting though.